src/prelude/env.savi

```diff
@@ -20,19 +20,55 @@
       i += 1
     )
 
-    @vars = []
+    @vars = EnvVars._from_envp(envp)
+
+  :fun "exit_code="(value)
+    LibPony.pony_exitcode(value)
+
+:class val EnvVars
+  :let _vars Array(EnvVar)
+
+  :new val _from_envp(envp CPointer(CPointer(U8)'ref)'ref)
+    @_vars = []
     if (envp.is_not_null) (
       while (True) (
         arg = envp._get_at(0)
         break if arg.is_null
-        len = InspectLibC.strlen(arg).usize
-        @vars.push(String.val_from_cpointer(arg, len, len))
+
+        try @_vars.push(EnvVar._from_cstring!(arg))
+
         envp = envp._offset(1)
       )
     )
 
-  :fun "exit_code="(value)
-    LibPony.pony_exitcode(value)
+  :fun each_pair
+    :yields (String'val, String'val) for None
+    @_vars.each_with_index -> (pair, _i |
+      yield (pair.key, pair.value)
+    )
+
+  :fun "[]"(needle String) (String | None)
+    @each_pair -> (key, value |
+      return value if (needle == key)
+    )
+    None
+
+  :fun "[]!"(needle String) String
+    @[needle].as!(String)
+
+:class val EnvVar
```

Comment:
Using `struct` instead of `class` here will result in less allocation overhead and less pointer indirection.

```suggestion
:struct val EnvVar
```

Note that this is not the same as Pony's struct. Broadly speaking, a Savi `:struct` is like a Pony tuple, except that it is named, its fields are named, and it can have methods. But the allocation semantics are the same as a Pony tuple.

Defining as a `:class` allocates a separate object on the heap for each instance, and makes it so that `Array(EnvVar)` contains the pointers to those `EnvVar` object instances.

Defining as a `:struct` allocates each one as a C-like struct value on the stack, and makes it so that `Array(EnvVar)` is allocated as if it were `Array(String)` of twice the length.

The use of `:struct` is allowed here because only `:let` fields are present in the type.

Reply:
When I change `:class` to `:struct` I get the following error trace:

```
[... snip ~11000 lines ...]
Savi compiler error occured with message "Module validation failed: Invalid bitcast
  %ARRAY.TRACE.LOOP.ELEM.OPAQUE = bitcast %EnvVar.FIELDS %ARRAY.TRACE.LOOP.ELEM to %_.OBJECT*
```

Reply:
Okay, thanks for trying. I'll investigate that struct issue separately.